### PR TITLE
Enhance side pane toggle

### PR DESCRIPTION
**Pull request by Braunch**
I improved the SidePanes toggle functionality. As described by @simonharrer at https://github.com/JabRef/jabref/issues/1225#issuecomment-222481080 the components should be shown, get focused if shown but not focused and closed when shown and focused.
Done for:
- [x] GroupSelector
- [x] GeneralFetcher 
- [x] OpenOfficePanel

Bonus:
- [ ] EntryEditor/PreviewPanel

Update:
The panes extending the SidePaneComponent have implemented toggle functionality. To finish the task i depend on @mairdl 's PR: https://github.com/JabRef/jabref/pull/1525 and his work on the group functionality. To make the toggle visible on the groupsSelector, the tree needs to be focusable and @mairdl is implementing it. Also the OOPanel depends on having a hotkey since the three phase toggling does not make sense when toggling using the mouse (you could click the panel right away instead of the menu item). So this PR will be on hold (hopefully) just for some days.

### CODE REVIEW

**Comment by Braunch** [comment]
@simonharrer how should the toggle for the preview panel and the entry editor work in your opinion?
Since the preview has nothing to focus the three phase toggle is not suitable here. For the entry editor on the other hand, I can imagine using the three phases. Any opinions on that (from anyone of course, just refering to @simonharrer because it was him to mention the idea in the first place)?

**Comment by koppor** [comment]
#1525 is merged. Will there be updates on this PR?

**Comment by bartsch-dev** [comment]
I took this issue to myself and coded a quick solution.

File: src/main/java/net/sf/jabref/gui/SidePaneManager.java [hunk 1]
```diff
@@ -67,6 +66,11 @@ public synchronized boolean isComponentVisible(String name) {
         }
     }
 
+    /**
+     * if panel is visible it will be hidden and the other way around
```
Comment (tschechlovdev) [inline src/main/java/net/sf/jabref/gui/SidePaneManager.java]: Please start comments with Capital letters.

File: src/main/java/net/sf/jabref/gui/SidePaneManager.java [hunk 2]
```diff
@@ -75,6 +79,21 @@ public synchronized void toggle(String name) {
         }
     }
 
+    /**
+     * if panel is hidden it will be shown and focused
```
Comment (tschechlovdev) [inline src/main/java/net/sf/jabref/gui/SidePaneManager.java]: Same as above.

File: src/main/java/net/sf/jabref/gui/SidePaneManager.java [hunk 3]
```diff
@@ -75,6 +79,21 @@ public synchronized void toggle(String name) {
...
+     * if panel is visible but not focused it will be focused
+     * ig panel is visible and focused it will be hidden
```
Comment (tschechlovdev) [inline src/main/java/net/sf/jabref/gui/SidePaneManager.java]: Typo ig -> if

File: src/main/java/net/sf/jabref/gui/SidePaneManager.java [hunk 4]
```diff
@@ -75,6 +79,21 @@ public synchronized void toggle(String name) {
...
+     *
+     * @param name name of the panel
+     */
+    public synchronized void toggleThreeWay(String name){
+        if (isComponentVisible(name) && Globals.getFocusListener().getFocused() == components.get(name)) {
```
Comment (tschechlovdev) [inline src/main/java/net/sf/jabref/gui/SidePaneManager.java]: Would be nice if you could add additional brackets here, that would make it clearer.

File: src/main/java/net/sf/jabref/gui/groups/GroupSelector.java [hunk 5]
```diff
@@ -702,6 +712,11 @@ public void componentOpening() {
     }
 
     @Override
+    public String getSidePaneName() {
```
Comment (tschechlovdev) [inline src/main/java/net/sf/jabref/gui/groups/GroupSelector.java]: Hm why do you set ``NAME`` as ``public static final`` and add this get method? Isn't this unnecessary then?

File: src/main/java/net/sf/jabref/gui/SidePaneManager.java (hunk 1 above)
Comment (bartsch-dev) [inline src/main/java/net/sf/jabref/gui/SidePaneManager.java]: done

File: src/main/java/net/sf/jabref/gui/SidePaneManager.java (hunk 2 above)
Comment (bartsch-dev) [inline src/main/java/net/sf/jabref/gui/SidePaneManager.java]: done

File: src/main/java/net/sf/jabref/gui/SidePaneManager.java (hunk 3 above)
Comment (bartsch-dev) [inline src/main/java/net/sf/jabref/gui/SidePaneManager.java]: done

File: src/main/java/net/sf/jabref/gui/SidePaneManager.java (hunk 4 above)
Comment (bartsch-dev) [inline src/main/java/net/sf/jabref/gui/SidePaneManager.java]: I extracted the second condition

**Comment by koppor** [comment]
`ig` should be `if`

File: src/main/java/net/sf/jabref/gui/groups/GroupSelector.java (hunk 5 above)
Comment (bartsch-dev) [inline src/main/java/net/sf/jabref/gui/groups/GroupSelector.java]: I need to access the name from the `SidePaneComponent`.
Sadly you cannot define an abstract static method

**Comment by koppor** [comment]
Refs https://github.com/koppor/jabref/issues/54

**Comment by koppor** [comment]
I tried it: Works good. Code LGTM.

File: src/main/java/net/sf/jabref/gui/JabRefFrame.java [hunk 6]
```diff
@@ -1519,8 +1504,8 @@ private void initActions() {
         openDatabaseOnlyActions.addAll(Arrays.asList(mergeDatabaseAction, newSubDatabaseAction, save, globalSearch,
                 saveAs, saveSelectedAs, saveSelectedAsPlain, editModeAction, undo, redo, cut, deleteEntry, copy, paste, mark, markSpecific, unmark,
                 unmarkAll, rankSubMenu, editEntry, selectAll, copyKey, copyCiteKey, copyKeyAndTitle, copyKeyAndLink, editPreamble, editStrings,
-                toggleGroups, makeKeyAction, normalSearch, generalFetcher.getAction(), mergeEntries, cleanupEntries, exportToClipboard, replaceAll,
-                sendAsEmail, downloadFullText, writeXmpAction, optMenuItem, findUnlinkedFiles, addToGroup, removeFromGroup,
+                groupSelector.getAction(), makeKeyAction, normalSearch, generalFetcher.getAction(), mergeEntries, cleanupEntries, exportToClipboard, replaceAll,
```
Comment (tobiasdiez) [inline src/main/java/net/sf/jabref/gui/JabRefFrame.java]: Maybe rename getAction -> getToggleAction ?

File: src/main/java/net/sf/jabref/gui/groups/GroupSelector.java [hunk 7]
```diff
@@ -1261,4 +1276,14 @@ public GroupsTree getGroupsTree() {
     public void listen(GroupUpdatedEvent updateEvent) {
         setGroups(updateEvent.getMetaData().getGroups().orElse(null));
     }
+
+    @Override
+    public void grabFocus() {
+        groupsTree.grabFocus();
+    }
+
+    public ToggleAction getAction() {
```
Comment (tobiasdiez) [inline src/main/java/net/sf/jabref/gui/groups/GroupSelector.java]: Make `getAction` an abstract method in SidePaneComponent ?

File: src/main/java/net/sf/jabref/gui/openoffice/OpenOfficePanel.java [hunk 8]
```diff
@@ -806,24 +782,27 @@ private void showSettingsPopup() {
...
     }
 
-    public String getName() {
-        return "OpenOffice/LibreOffice";
+    public SidePaneComponent.ToggleAction getAction() {
+        return sidePane.getToggleAction();
     }
 
 
     private class OOPanel extends SidePaneComponent {
```
Comment (tobiasdiez) [inline src/main/java/net/sf/jabref/gui/openoffice/OpenOfficePanel.java]: Can you please extract this private class to a new file. Thanks.

Comment (bartsch-dev) [inline src/main/java/net/sf/jabref/gui/openoffice/OpenOfficePanel.java]: done

File: src/main/java/net/sf/jabref/gui/groups/GroupSelector.java (hunk 7 above)
Comment (bartsch-dev) [inline src/main/java/net/sf/jabref/gui/groups/GroupSelector.java]: Done, although `FileUpdatePanel` doesn't implements it.

File: src/main/java/net/sf/jabref/gui/JabRefFrame.java (hunk 6 above)
Comment (bartsch-dev) [inline src/main/java/net/sf/jabref/gui/JabRefFrame.java]: done

**Comment by bartsch-dev** [comment]
When a subclass does __NOT__ implement an abstract method, how should I mark it (look at https://github.com/JabRef/jabref/pull/1605/files#diff-742e26b1d3d71ccf826992589b1b666cR77)?
Throwing a `NotImplementedException` breaks the style check.

**Comment by tobiasdiez** [comment]
Code looks good to me now (thanks for the quick changes!). I'll merge it in.

**Pull request merged**